macro/src/lib.rs

```diff
@@ -139,20 +149,65 @@ pub fn subxt(args: TokenStream, input: TokenStream) -> TokenStream {
             .collect(),
     );
     for d in args.attributes_for_type {
+        validate_type_path(&d.path.path, &metadata);
         codegen.add_attributes_for_type(d.path, d.attributes.into_iter().map(|a| a.0), d.recursive)
     }
 
     // Insert type substitutions:
     for sub in args.substitute_type.into_iter() {
+        validate_type_path(&sub.path, &metadata);
         codegen.set_type_substitute(sub.path, sub.with);
     }
 
+    let code = codegen
+        .generate(metadata)
+        .map_err(|e| e.into_compile_error())?;
+
+    Ok(code.into())
+}
+
+/// Checks that a type is present in the type registry. If it is not found, abort with a
+/// helpful error message, showing the user alternative types, that have the same name, but are at different locations in the metadata.
+fn validate_type_path(path: &syn::Path, metadata: &Metadata) {
+    let path_segments = path_segments(path);
+    let ident = &path
+        .segments
+        .last()
+        .expect("Empty path should be filtered out before already")
+        .ident;
+    if !registry_contains_type_path(metadata.types(), &path_segments) {
+        let alternatives = similar_type_paths_in_registry(metadata.types(), path);
+        let alternatives: String = if alternatives.is_empty() {
+            format!("There is no Type with name `{ident}` in the provided metadata.")
+        } else {
+            let mut s = "A type with the same name is present at: ".to_owned();
+            for p in alternatives {
+                s.push('\n');
+                s.push_str(&pretty_path(&p));
+            }
+            s
+        };
+
+        abort_call_site!(
+            "Type `{}` does not exist at path `{}`\n\n{}",
+            ident.to_string(),
+            pretty_path(path),
+            alternatives
+        );
+    }
+
+    fn pretty_path(path: &syn::Path) -> String {
+        path.to_token_stream().to_string().replace(' ', "")
+    }
+}
+
+/// Fetches metadata in a blocking manner, either from a url (not recommended) or from a file path.
```

Comment:
I don't get why fetching metadata from an URL is not recommended? :)

Reply:
It is not recommended for users because the compilation blocks while fetching the metadata, which is a bit slow and (maybe flaky) compared to using metadata file. But yeah, maybe the docs of this function don't need to mention this.

Reply:
Maybe we can say its not recommended for production use-cases. Do remember we added the ability to fetch the metadata from URL as a feature request from the `cargo-contract` team, that wanted to fetch a fresh metadata for their testing